File: ShiftScheduler.java
import java.util.*;

class Employee {
    String name;
    List<String> preferences;
    int daysWorked;

    Employee(String name, List<String> preferences) {
        this.name = name;
        this.preferences = preferences;
        this.daysWorked = 0;
    }
}

public class ShiftScheduler {
    static String[] DAYS = {"Monday","Tuesday","Wednesday","Thursday","Friday"};
    static String[] SHIFTS = {"Morning","Afternoon","Evening"};

    public static void main(String[] args) {
        Map<String, Employee> employees = new HashMap<>();
        employees.put("Alice", new Employee("Alice", Arrays.asList("Morning")));
        employees.put("Bob", new Employee("Bob", Arrays.asList("Afternoon")));
        employees.put("Charlie", new Employee("Charlie", Arrays.asList("Evening")));
        employees.put("David", new Employee("David", Arrays.asList("Morning","Afternoon")));
        employees.put("Ella", new Employee("Ella", Arrays.asList("Afternoon","Evening")));
        employees.put("Frank", new Employee("Frank", Arrays.asList("Morning","Evening")));

        Map<String, Map<String, List<String>>> schedule = new LinkedHashMap<>();
        for (String day : DAYS) {
            Map<String, List<String>> shifts = new LinkedHashMap<>();
            for (String shift : SHIFTS) shifts.put(shift, new ArrayList<>());
            schedule.put(day, shifts);
        }

        Random rand = new Random();
        for (String day : DAYS) {
            Set<String> assignedToday = new HashSet<>();
            for (String shift : SHIFTS) {
                List<String> available = new ArrayList<>();
                for (Employee e : employees.values()) {
                    if (e.preferences.contains(shift) && e.daysWorked < 5 && !assignedToday.contains(e.name)) {
                        available.add(e.name);
                    }
                }
                Collections.shuffle(available);
                while (schedule.get(day).get(shift).size() < 2) {
                    String emp = null;
                    if (!available.isEmpty()) emp = available.remove(0);
                    else {
                        List<String> candidates = new ArrayList<>();
                        for (Employee e : employees.values()) {
                            if (e.daysWorked < 5 && !assignedToday.contains(e.name))
                                candidates.add(e.name);
                        }
                        if (candidates.isEmpty()) break;
                        emp = candidates.get(rand.nextInt(candidates.size()));
                    }
                    schedule.get(day).get(shift).add(emp);
                    employees.get(emp).daysWorked++;
                    assignedToday.add(emp);
                }
            }
        }

        System.out.println("=== Weekly Schedule ===");
        for (String day : DAYS) {
            System.out.println("\n" + day);
            for (String shift : SHIFTS) {
                System.out.println("  " + shift + ": " + String.join(", ", schedule.get(day).get(shift)));
            }
        }
    }
}

File: shift_scheduler.py
import random

DAYS = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday"]
SHIFTS = ["Morning", "Afternoon", "Evening"]

employees = {
    "Alice": {"preferences": ["Morning"], "days_worked": 0},
    "Bob": {"preferences": ["Afternoon"], "days_worked": 0},
    "Charlie": {"preferences": ["Evening"], "days_worked": 0},
    "David": {"preferences": ["Morning", "Afternoon"], "days_worked": 0},
    "Ella": {"preferences": ["Afternoon", "Evening"], "days_worked": 0},
    "Frank": {"preferences": ["Morning", "Evening"], "days_worked": 0},
}

schedule = {day: {shift: [] for shift in SHIFTS} for day in DAYS}

for day in DAYS:
    assigned_today = set()
    for shift in SHIFTS:
        available = [
            name for name, data in employees.items()
            if shift in data["preferences"]
            and data["days_worked"] < 5
            and name not in assigned_today
        ]
        random.shuffle(available)
        while len(schedule[day][shift]) < 2:
            if available:
                emp = available.pop()
            else:
                candidates = [
                    n for n, d in employees.items()
                    if d["days_worked"] < 5 and n not in assigned_today
                ]
                if not candidates:
                    break
                emp = random.choice(candidates)
            schedule[day][shift].append(emp)
            employees[emp]["days_worked"] += 1
            assigned_today.add(emp)

print("=== Weekly Schedule ===")
for day, shifts in schedule.items():
    print(f"\n{day}")
    for shift, names in shifts.items():
        print(f"  {shift}: {', '.join(names)}")
